File: src/IsbInterop/Data/Proxies/Requisite.vb
Imports IsbInterop.DataTypes.Enumerable

Namespace Data.Proxies

  ''' <summary>
  ''' Обертка над IRequisite.
  ''' </summary>
  Friend Class Requisite
    Inherits BaseIsbObject
    Implements IRequisite

#Region "Поля и свойства"

    ''' <summary>
    ''' Значение реквизита в качестве даты.
    ''' </summary>
    Public ReadOnly Property AsDate As DateTime Implements IRequisite.AsDate
      Get
        Return GetRcwProperty("AsDate")
      End Get
    End Property

    ''' <summary>
    ''' Целочисленное значение реквизита.
    ''' </summary>
    Public ReadOnly Property AsInteger As Integer Implements IRequisite.AsInteger
      Get
        Return GetRcwProperty("AsInteger")
      End Get
    End Property

    ''' <summary>
    ''' Строковое значение реквизита.
    ''' </summary>
    Public ReadOnly Property AsString As String Implements IRequisite.AsString
      Get
        Return DirectCast(GetRcwProperty("AsString"), String)
      End Get
    End Property

    ''' <summary>
    ''' Тип реквизита.
    ''' </summary>
    Public ReadOnly Property DataType As TReqDataType Implements IRequisite.DataType
      Get
        Dim requisiteDataType = CInt(GetRcwProperty("DataType"))

        If [Enum].IsDefined(GetType(TReqDataType), requisiteDataType) Then
          Return requisiteDataType
        Else
          Return TReqDataType.rdtUnknown
        End If
      End Get
    End Property

    ''' <summary>
    ''' Имя поля реквизита в базе.
    ''' </summary>
    Public ReadOnly Property FieldName As String Implements IRequisite.FieldName
      Get
        Return DirectCast(GetRcwProperty("FieldName"), String)
      End Get
    End Property

    ''' <summary>
    ''' Признак, указывающий, что реквизит равен null.
    ''' </summary>
    Public ReadOnly Property IsNull As Boolean Implements IRequisite.IsNull
      Get
        Return GetRcwProperty("IsNull")
      End Get
    End Property

    ''' <summary>
    ''' Имя реквизита.
    ''' </summary>
    Public ReadOnly Property Name As String Implements IRequisite.Name
      Get
        Return DirectCast(GetRcwProperty("Name"), String)
      End Get
    End Property

    ''' <summary>
    ''' Значение реквизита.
    ''' </summary>
    Public Property Value As Object Implements IRequisite.Value
      Get
        Return GetRcwProperty("Value")
      End Get
      Set(requisiteValue As Object)
        SetRcwProperty("Value", requisiteValue)
      End Set
    End Property

#End Region

#Region "Конструктор"

    ''' <summary>
    ''' Конструктор.
    ''' </summary>
    ''' <param name="rcwRequisite">COM-объект реквизит.</param>
    ''' <param name="scope">Область видимости.</param>
    Friend Sub New(rcwRequisite As Object, scope As IScope)
      MyBase.New(rcwRequisite, scope)
    End Sub

#End Region

  End Class
End Namespace

File: src/IsbInterop/Data/Proxies/Requisites/Requisite.cs
using IsbInterop.DataTypes.Enumerable;
using System;
using IsbInterop.Data.Requisites;

namespace IsbInterop.Data.Proxies.Requisites
{
  /// <summary>
  /// Обертка над IRequisite.
  /// </summary>
  internal class Requisite : BaseIsbObject, IRequisite
  {
    #region Поля и свойства

    /// <summary>
    /// Значение реквизита в качестве даты.
    /// </summary>
    public DateTime AsDate => (DateTime)GetRcwProperty("AsDate");

    /// <summary>
    /// Целочисленное значение реквизита.
    /// </summary>
    public int AsInteger => (int)GetRcwProperty("AsInteger");

    /// <summary>
    /// Строковое значение реквизита.
    /// </summary>
    public string AsString => (string)GetRcwProperty("AsString");

    /// <summary>
    /// Тип реквизита.
    /// </summary>
    public TReqDataType DataType
    {
      get
      {
        int dataType = (int)GetRcwProperty("DataType");

        if (Enum.IsDefined(typeof(TReqDataType), dataType))
          return (TReqDataType)dataType;
        else
          return TReqDataType.rdtUnknown;
      }
    }

    /// <summary>
    /// Имя поля реквизита в базе.
    /// </summary>
    public string FieldName => (string)GetRcwProperty("FieldName");

    /// <summary>
    /// Признак, указывающий, что реквизит равен null.
    /// </summary>
    public bool IsNull => (bool)GetRcwProperty("IsNull");

    /// <summary>
    /// Имя реквизита.
    /// </summary>
    public string Name => (string)GetRcwProperty("Name");

    /// <summary>
    /// Значение реквизита.
    /// </summary>
    public object Value
    {
      get { return GetRcwProperty("Value"); }
      set { SetRcwProperty("Value", value); }
    }

    #endregion

    #region Конструктор

    /// <summary>
    /// Конструктор.
    /// </summary>
    /// <param name="rcwRequisite">COM-объект реквизит.</param>
    /// <param name="scope"></param>
    internal Requisite(object rcwRequisite, IScope scope) : base(rcwRequisite, scope) { }

    #endregion
  }
}
